File: HYAwesomeTransitionDemo/ViewController.m
//
//  ViewController.m
//  HYAwesomeTransitionDemo
//
//  Created by nathan on 15/7/30.
//  Copyright (c) 2015年 nathan. All rights reserved.
//

#import "ViewController.h"
#import "HYAwesomeTransition.h"
#import "ModalViewController.h"
#import "CustomCell.h"


@interface ViewController ()<UICollectionViewDataSource,UICollectionViewDelegate,UIViewControllerTransitioningDelegate,ModalViewControllerDelegate,UINavigationControllerDelegate>
@property (nonatomic, weak) IBOutlet UICollectionView *collectionView;
@property (nonatomic, strong) HYAwesomeTransition *awesometransition;
@property (nonatomic, strong) NSMutableArray *imagesArray;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.imagesArray = @[].mutableCopy;
    for (int i = 0; i < 30; i++) {
        [self.imagesArray addObject:@"doge2"];
    }
    [self.imagesArray replaceObjectAtIndex:10 withObject:@"doge"];
    
    self.awesometransition = [[HYAwesomeTransition alloc] init];
    self.awesometransition.type = HYTransitionTypeModal;
    self.awesometransition.duration = 1.5f;
    self.awesometransition.containerBackgroundView = ({
        UIView *bgView = (UIView *)[[[NSBundle mainBundle] loadNibNamed:@"ContainerBackgroundView" owner:nil options:nil] lastObject];
        bgView;
    });
}

#pragma mark - collectionView


- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *ReuseIdentifier = @"CustomMainCell";
    CustomCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:ReuseIdentifier forIndexPath:indexPath];
    
    cell.imageView.image = [UIImage imageNamed:self.imagesArray[indexPath.row]];
    
    return cell;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.imagesArray.count;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    UICollectionViewCell *cell = [collectionView cellForItemAtIndexPath:indexPath];
    
    if (self.awesometransition.type == HYTransitionTypeModal) {
        
        ModalViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"ModalViewController"];
        vc.imageName = self.imagesArray[indexPath.row];
        vc.transitioningDelegate = self;
        vc.delegate              = self;
        
        CGRect startFrame = [cell convertRect:cell.bounds toView:self.view];
        CGRect finalFrame = CGRectMake(40, 150, 100, 100);
        
        [self.awesometransition registerStartFrame:startFrame
                                        finalFrame:finalFrame transitionView:cell];
        
        [self presentViewController:vc animated:YES completion:^{
            vc.avatar.hidden = NO;
            vc.avatar.image = [UIImage imageNamed:self.imagesArray[indexPath.row]];
        }];
    } else {
        ModalViewController *vc = [self.storyboard instantiateViewControllerWithIdentifier:@"ModalViewController"];
        vc.imageName = self.imagesArray[indexPath.row];
        vc.delegate              = self;
        
        CGRect startFrame = [cell convertRect:cell.bounds toView:self.view];
        CGRect finalFrame = CGRectMake(40, 150, 100, 100);
        
        [self.awesometransition registerStartFrame:startFrame
                                        finalFrame:finalFrame transitionView:cell];
        
        self.navigationController.delegate = self;
        
        [self.navigationController pushViewController:vc animated:YES];
    }
}

#pragma mark - ModelViewController delegate

- (void)modalViewControllerDidClickedDismissButton:(ModalViewController *)viewController {
    self.awesometransition.finalFrame = [viewController.avatar convertRect:viewController.avatar.bounds toView:viewController.view];
    viewController.avatar.hidden = YES;
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma makr - Action

- (IBAction)segmentedControlAction:(UISegmentedControl *)sender {
    if (sender.selectedSegmentIndex == 1) {
        self.awesometransition.type = HYTransitionTypeNavigation;
    } else {
        self.awesometransition.type = HYTransitionTypeModal;
    }
}

#pragma mark - transition

- (id <UIViewControllerAnimatedTransitioning>)animationControllerForPresentedController:(UIViewController *)presented presentingController:(UIViewController *)presenting sourceController:(UIViewController *)source {
    self.awesometransition.present = YES;
    return self.awesometransition;
}

- (id<UIViewControllerAnimatedTransitioning>)animationControllerForDismissedController:(UIViewController *)dismissed {
    self.awesometransition.present = NO;
    return self.awesometransition;
}

// If you use UINavigationController, you have to implement UINavigationControllerDelegate
// instead of UIViewControllerTransitioningDelegate

- (id<UIViewControllerAnimatedTransitioning>)navigationController:(UINavigationController *)navigationController animationControllerForOperation:(UINavigationControllerOperation)operation fromViewController:(UIViewController *)fromVC toViewController:(UIViewController *)toVC{
    
    self.awesometransition.type = HYTransitionTypeNavigation;
    if (operation == UINavigationControllerOperationPush) {
        self.awesometransition.present = YES;
    } else {
        self.awesometransition.present = NO;
    }
    return self.awesometransition;
}


@end

File: AwesomeTransition/AwesomeTransition/ViewController.swift
//
//  ViewController.swift
//  AwesomeTransition
//
//  Created by why on 16/4/18.
//  Copyright © 2016年 why. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

	@IBOutlet weak var collectionView: UICollectionView!
    var imageStrings: [String] = []
    lazy var transition: AwesomeTransition = {
        let transition = AwesomeTransition()
        transition.containerBackgroundView = Bundle.main.loadNibNamed("ContainerView", owner: nil, options: nil)?.last as? UIView
        return transition
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        for _ in 0...30 {
            imageStrings.append("doge")
        }
        imageStrings[10] = "doge2"
    }
}

extension ViewController: UICollectionViewDataSource, UICollectionViewDelegate {
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return imageStrings.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CustomCell", for: indexPath) as! CustomCell
        cell.imageView.image = UIImage(named: imageStrings[indexPath.item])
        return cell
    }
    
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        
        let cell = collectionView.cellForItem(at: indexPath)
        
        let detailController = self.storyboard?.instantiateViewController(withIdentifier: "detailController") as? DetailViewController
        
        if let detailController = detailController, let cell = cell{
            detailController.imageName = imageStrings[indexPath.item]
            detailController.transitioningDelegate = self
            detailController.delegate = self
            
            let startFrame = cell.convert(cell.bounds, to: self.view)
            let finalFrame = CGRect(x: 40, y: 170, width: 100, height: 100)
            
            self.transition.registerTransition(startFrame, finalRect: finalFrame, transitionView: cell)
            
            self.present(detailController, animated: true, completion: { 
                detailController.imageView.isHidden = false
                detailController.imageView.image = UIImage(named: self.imageStrings[indexPath.item])
            })
        }
        
    }
}

extension ViewController: DetailViewControllerDelegate {
    func detailViewController(_ detailViewController: DetailViewController, didDismiss: Bool) {
        transition.finalFrame = detailViewController.imageView.convert(detailViewController.imageView.bounds, to: detailViewController.view)
        detailViewController.imageView.isHidden = true
        self.dismiss(animated: true, completion: nil)
    }
}

extension ViewController: UIViewControllerTransitioningDelegate {
    
    func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        transition.isPresent = true
        return transition
    }
    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        transition.isPresent = false
        return transition
    }
    
}

//extension ViewController: UINavigationControllerDelegate {
//    
//    func navigationController(navigationController: UINavigationController, animationControllerForOperation operation: UINavigationControllerOperation, fromViewController fromVC: UIViewController, toViewController toVC: UIViewController) -> UIViewControllerAnimatedTransitioning? {
//        
//    }
//}
